satpy/composites/__init__.py

```diff
@@ -549,71 +549,111 @@ def _insert_palette_colors(channels, palette):
         try:
-            coszen = np.cos(np.deg2rad(projectables[2]))
+            coszen = np.cos(np.deg2rad(projectables[2 if self.day_night == "day_night" else 1]))
         except IndexError:
             from pyorbital.astronomy import cos_zen
             LOG.debug("Computing sun zenith angles.")
             # Get chunking that matches the data
             try:
-                chunks = day_data.sel(bands=day_data['bands'][0]).chunks
+                chunks = foreground_data.sel(bands=foreground_data['bands'][0]).chunks
             except KeyError:
-                chunks = day_data.chunks
-            lons, lats = day_data.attrs["area"].get_lonlats(chunks=chunks)
-            coszen = xr.DataArray(cos_zen(day_data.attrs["start_time"],
+                chunks = foreground_data.chunks
+            lons, lats = foreground_data.attrs["area"].get_lonlats(chunks=chunks)
+            coszen = xr.DataArray(cos_zen(foreground_data.attrs["start_time"],
                                           lons, lats),
                                   dims=['y', 'x'],
-                                  coords=[day_data['y'], day_data['x']])
+                                  coords=[foreground_data['y'], foreground_data['x']])
         # Calculate blending weights
         coszen -= np.min((lim_high, lim_low))
         coszen /= np.abs(lim_low - lim_high)
         coszen = coszen.clip(0, 1)
 
         # Apply enhancements to get images
-        day_data = enhance2dataset(day_data)
-        night_data = enhance2dataset(night_data)
+        foreground_data = enhance2dataset(foreground_data)
+        background_data = enhance2dataset(background_data)
+
+        # Add an alpha band for L or RGB mode
+        new_foreground_data = [foreground_data.sel(bands=band) for band in foreground_data['bands'].data]
+        new_background_data = [background_data.sel(bands=band) for band in background_data['bands'].data]
+        foreground_alpha = new_foreground_data[0].copy()
+        background_alpha = new_background_data[0].copy()
+        foreground_alpha.data = da.ones((foreground_data.sizes['y'],
+                                         foreground_data.sizes['x']),
+                                        chunks=new_foreground_data[0].chunks)
+        background_alpha.data = da.ones((background_data.sizes['y'],
+                                         background_data.sizes['x']),
+                                        chunks=new_background_data[0].chunks)
+        foreground_alpha['bands'] = 'A'
+        background_alpha['bands'] = 'A'
+        new_foreground_data.append(foreground_alpha)
+        new_background_data.append(background_alpha)
+        new_foreground_data = xr.concat(new_foreground_data, dim='bands')
+        new_background_data = xr.concat(new_background_data, dim='bands')
+        new_foreground_data.attrs['mode'] = foreground_data.attrs['mode'] + 'A'
+        new_background_data.attrs['mode'] = background_data.attrs['mode'] + 'A'
```

Comment:
The lines here are duplicated, we should have a new function for this and call it respectively on foreground and background data.

Reply:
OK, I've made a new one `add_alpha_bands`.